python/lsst/obs/base/repodb/base.py

```diff
@@ -0,0 +1,442 @@
+        `UnitMeta`). Putting all of this logic in `ForeignKey` avoids circular
+        dependency problems.
+
+        Parameters
+        ----------
+        cls : `type` derived from `Unit`
+            `Unit` class to which this field has been added.
+        name : `str`
+            Name of the class attribute used to store this field; will be used
+            to set the `Field`'s internal name attribute'
+        """
+        Field.attach(self, cls, name)
+        if self._reverse is not None:
+            reverse = getattr(self.UnitClass, self._reverse)
+            reverse.attach(self.UnitClass, self._reverse)
+
+    def finalize(self, instance, others):
+        """
+        Complete initialization of a `Unit` instance holding this field.
+
+        Should only be called by `Backend` implementations when constructing a
+        `Unit` instance from SQL query results.
+
+        When being constructed by a `Backend` from a SQL query, a
+        `ForeignKey`'s storage value may be set to the integer ID of the
+        foreign `Unit` instance, rather than the instance itself, and
+        `ReverseForeignKey` `sets` may be left unfilled.  When this is the
+        case, `ForeignKey.finalize` may be called to transform the integer ID
+        values to `Unit` instances and populate the `ReverseForeignKey` sets.
+        """
+        id = instance._storage.get(self.name, None)
+        if id is None:
+            assert self.optional
+            return
+        target = others[self.UnitClass][id]
+        instance._storage[self.name] = target
+        rev = target._reversed.setdefault(self._reverse, set())
+        rev.add(instance)
+
+
+class ReverseForeignKey(object):
+    """A descriptor for `Unit` classes that captures a one-to-many
+    relationship.
+
+    A `ReverseForeignKey` must be paired with a `ForeignKey` class in the
+    `Unit` subclass to which it refers, and is not a true `Field` becaues it
+    relies on the `ForeignKey` for its initialization and database storage.
+
+    `ReverseForeignKey`s values are stored in a `_reversed` dictionary
+    attribute on the `Unit` instance that holds it, and are typically populated
+    by `ForeignKey.finalize`.
+
+    `ReverseForeignKey`s may be empty and unusable when the database query that
+    generates the `Unit`s does not extend to the `Unit` type referred to by
+    the `ReverseForeignKey` (queries are always expanded to include the `Unit`s
+    referred to by a `ForeignKey`, but this behavior is deliberately asymettric
+    because a `Unit`'s `ReverseForeignKey` values cannot be used to uniquely
+    define it).'
+    """
+
+    def __init__(self):
+        self.name = None
+
+    def __get__(self, instance, owner=None):
+        if instance is not None:
+            return instance._reversed.get(self.name, None)
+        return self
+
+    def attach(self, cls, name):
+        """Attach this descriptor to a `Unit` type.
+
+        Should only be called by `ForeignKey.attach`.
```

Comment:
Good catch; that was a mistake that I've now fixed.  I think it worked before because the incorrect `attach` call from `UnitMeta` was overridden by the correct one by `ForeignKey.attach`.